chain/src/consume_unverified.rs

```diff
@@ -1,405 +1,245 @@
-        }
-    }
-}
+    unverified_block_rx: Receiver<UnverifiedBlock>,
+    truncate_block_rx: Receiver<TruncateRequest>,
 
-pub(crate) struct GlobalIndex {
-    pub(crate) number: BlockNumber,
-    pub(crate) hash: Byte32,
-    pub(crate) unseen: bool,
+    stop_rx: Receiver<()>,
+    processor: ConsumeUnverifiedBlockProcessor,
 }
 
-impl GlobalIndex {
-    pub(crate) fn new(number: BlockNumber, hash: Byte32, unseen: bool) -> GlobalIndex {
-        GlobalIndex {
-            number,
-            hash,
-            unseen,
+impl ConsumeUnverifiedBlocks {
+    pub(crate) fn new(
+        shared: Shared,
+        unverified_blocks_rx: Receiver<UnverifiedBlock>,
+        truncate_block_rx: Receiver<TruncateRequest>,
+        proposal_table: ProposalTable,
+        verify_failed_blocks_tx: tokio::sync::mpsc::UnboundedSender<VerifyFailedBlockInfo>,
+        stop_rx: Receiver<()>,
+    ) -> Self {
+        ConsumeUnverifiedBlocks {
+            tx_pool_controller: shared.tx_pool_controller().to_owned(),
+            unverified_block_rx: unverified_blocks_rx,
+            truncate_block_rx,
+            stop_rx,
+            processor: ConsumeUnverifiedBlockProcessor {
+                shared,
+                proposal_table,
+                verify_failed_blocks_tx,
+            },
         }
     }
 
-    pub(crate) fn forward(&mut self, hash: Byte32) {
-        self.number -= 1;
-        self.hash = hash;
-    }
-}
-
-/// Chain background service
-///
-/// The ChainService provides a single-threaded background executor.
-pub struct ChainService {
-    shared: Shared,
-    proposal_table: ProposalTable,
-}
-
-impl ChainService {
-    /// Create a new ChainService instance with shared and initial proposal_table.
-    pub fn new(shared: Shared, proposal_table: ProposalTable) -> ChainService {
-        ChainService {
-            shared,
-            proposal_table,
-        }
-    }
+    pub(crate) fn start(mut self) {
+        loop {
+            let _trace_begin_loop = minstant::Instant::now();
+            select! {
+                recv(self.unverified_block_rx) -> msg => match msg {
+                    Ok(unverified_task) => {
+                        // process this unverified block
+                        trace!("got an unverified block, wait cost: {:?}", begin_loop.elapsed());
+                        if let Some(handle) = ckb_metrics::handle() {
+                            handle.ckb_chain_consume_unverified_block_waiting_block_duration_sum.add(_trace_begin_loop.elapsed().as_secs_f64())
+                        }
+                        let _ = self.tx_pool_controller.suspend_chunk_process();
 
-    /// start background single-threaded service with specified thread_name.
-    pub fn start<S: ToString>(mut self, thread_name: Option<S>) -> ChainController {
-        let signal_receiver = new_crossbeam_exit_rx();
-        let (process_block_sender, process_block_receiver) = channel::bounded(DEFAULT_CHANNEL_SIZE);
```

Comment:
Since `consume_unverified_blocks` actually does two things, verifying the block and invoking the callback. It would be better if we can make two metrics instead of a single one.

Reply:
How about keep `ckb_chain_consume_unverified_block_duration_sum` and add another metric: `ckb_chain_execute_callback_duration_sum`.

Reply:
LGTM